### PR TITLE
[DRAFT] Declarative Port Graph

### PR DESCRIPTION
**Pull request by revol-xut**
# Declarative Port Graph

This pull request is a tremendous refactor of reactor-cpp instead of connecting ports in the assemble methods, this pull requests introduces a way to declarative define the port graph. On this more formal definition of a graph (`PropertyGraph`) I implemented a spanning tree style optimization that figures out all the connected downstream ports. Consequently, this spanning tree is then "flattened" into a tree of depth one in the process all the connections to every leaf of the tree is merged. This merging logic also takes Delayed, Enclaved and Physical connections into consideration. 

## Important Changes for the Code Generator

**Declaring the PortGraph**
```cpp
# pushing the port objects into the environment
auto add_i1 = env.register_port(&add.i1);
auto add_i2 = env.register_port(&add.i2);
auto add_sum = env.register_port(&add.sum);
auto p_add_forward = env.register_port(&p_add.forward);
auto p_add_value = env.register_port(&p_add.value);

# connecting up the ports with their identifier 
env.draw_connection(counter1_count, add_i1, ConnectionProperties{});
env.draw_connection(counter2_count, add_i2, ConnectionProperties{});
env.draw_connection(add_sum, p_add_forward, ConnectionProperties{ConnectionType::Delayed, 10s, nullptr});
env.draw_connection(p_add_forward, p_add_value, ConnectionProperties{ConnectionType::Delayed, 5s, nullptr});
```
**Start Up**
```cpp
# first we need to "optimize" the graph this is currently a mandetory 
env.optimize();

# assembeling everything
# 1.) assembeling reactors
# 2.) intantiating port graph
# 3.) calculating dependency graph and calculate reaction indices
env.assemble();

# beginning execution
auto thread = env.startup();
thread.join();
```

### CODE REVIEW

**Comment by lhstrh** [comment]
This looks interesting, but could you provide a high-level description of what this is meant to do?

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -14,33 +14,33 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-latest, windows-latest, macos-latest]
+        os: [ ubuntu-latest, windows-latest, macos-latest ]
```
Comment (cmnrd) [inline .github/workflows/main.yaml:17]: Why this change?

File: .github/workflows/main.yaml [hunk 2]
```diff
@@ -14,33 +14,33 @@ jobs:
...
-        cmake --build build --target examples
-    - name: build with tracing enabled
-      run: |
-        sudo apt-get install -y liblttng-ust-dev
-        cd build
-        cmake -DREACTOR_CPP_TRACE=ON ..
-        cd ..
-        cmake --build build --target examples
-      if: matrix.os == 'ubuntu-latest'
+      - uses: actions/checkout@v1
```
Comment (cmnrd) [inline .github/workflows/main.yaml:20]: Were these tabs?

File: CMakeLists.txt [hunk 3]
```diff
@@ -9,8 +9,8 @@ endif()
 
 project(${LIB_TARGET} LANGUAGES CXX VERSION 0.0.1)
 
-# require C++17
-set(CMAKE_CXX_STANDARD 17)
+# require C++20
```
Comment (cmnrd) [inline CMakeLists.txt:12]: I am hesitant to switch to C++20. Unfortunately, it is still not so widely supported. I recently tried to use some C++20 feature only to learn that they are not available in the CI runners. Is there a particular reason to change the C++ verstion?

File: include/reactor-cpp/scheduler.hh [hunk 4]
```diff
@@ -193,6 +193,16 @@ public:
   };
 
   void set_port(BasePort* port);
```
Comment (cmnrd) [inline include/reactor-cpp/scheduler.hh:195]: Is this still used in your code?

**Review by cmnrd — COMMENTED** [review]
I left several comments and questions. This is quite a tremendous refactoring, but I think there is still a lot of cleaning up needed and I would like to better understand the motivation behind some of the changes. . The most pressing task, however, is to adjust the LF code generator and see if our tests are passing. This should be done before we even consider any optimizations, because only with tests we can be sure that we don't break anything.

I also second Marten's request. Could you add a description of what this PR is doing on a high level and why?

File: include/reactor-cpp/connection.hh [hunk 5]
```diff
@@ -39,15 +41,20 @@ protected:
...
+  virtual auto upstream_set_callback() noexcept -> PortCallback = 0;
   virtual void bind_upstream_port(Port<T>* port) {
     reactor_assert(upstream_port_ == nullptr);
     upstream_port_ = port;
     port->register_set_callback(upstream_set_callback());
   }
 
+  virtual void bind_downstream_ports(const std::vector<BasePort*>& ports) {
+    // with C++23 we can use insert_rage here
+    for ([[maybe_unused]] auto* port : ports) { // NOLINT
```
[the remark this answers is not in the document]
Comment (revol-xut) [reply include/reactor-cpp/connection.hh:54]: because it complaints that port is unused

File: CMakeLists.txt (hunk 3 above)
Comment (revol-xut) [reply CMakeLists.txt:12]: I used `.contains` at some places. But I also can just do it with `.find`

File: .github/workflows/main.yaml (hunk 2 above)
Comment (revol-xut) [reply .github/workflows/main.yaml]: reverted back

File: CMakeLists.txt (hunk 3 above)
Comment (revol-xut) [reply CMakeLists.txt]: reverted back to 17

File: lib/environment.cc [hunk 6]
```diff
@@ -229,7 +326,7 @@ void Environment::calculate_indexes() {
     }
 
     if (degree_zero.empty()) {
-      export_dependency_graph("/tmp/reactor_dependency_graph.dot");
+      // export_dependency_graph("/tmp/reactor_dependency_graph.dot");
```
[the remark this answers is not in the document]
Comment (revol-xut) [reply lib/environment.cc:325]: I would suggest lets make a seperate pr where we add back the debugging methods

File: include/reactor-cpp/connection.hh [hunk 7]
```diff
@@ -21,6 +21,8 @@
 #include "time.hh"
 #include "time_barrier.hh"
 
+#define _U_ __attribute__((unused)) // NOLINT macro to surpress unused warning by clang
```
[the remark this answers is not in the document]
Comment (revol-xut) [reply include/reactor-cpp/connection.hh]: done

File: lib/scheduler.cc [hunk 8]
```diff
@@ -502,25 +502,7 @@ void Scheduler::set_port(BasePort* port) {
   set_ports_[Worker::current_worker_id()].push_back(port);
 
   // recursively search for triggered reactions
-  set_port_helper(port);
-}
-
-void Scheduler::set_port_helper(BasePort* port) {
```
[the remark this answers is not in the document]
Comment (revol-xut) [reply lib/scheduler.cc:508]: because there are now more efficant methods for it

File: include/reactor-cpp/port.hh [hunk 9]
```diff
@@ -73,35 +85,38 @@ protected:
...
+  inline void set_tag(std::size_t tag) noexcept { tag_ = tag; }
+
+  virtual void pull_connection(const ConnectionProperties& properties, const std::vector<BasePort*>& downstreams) = 0;
 
   void register_set_callback(const PortCallback& callback);
   void register_clean_callback(const PortCallback& callback);
 
   friend class Reaction;
   friend class Scheduler;
+  virtual void set_inward_binding(BasePort* port) noexcept = 0;
```
[the remark this answers is not in the document]
Comment (revol-xut) [reply include/reactor-cpp/port.hh:109]: sadly we cannot because this function is used in `environment.cc` where we dont have anytype information

File: include/reactor-cpp/port.hh [hunk 10]
```diff
@@ -40,11 +44,19 @@ protected:
 
   BasePort(const std::string& name, PortType type, Reactor* container)
       : ReactorElement(name, match_port_enum(type), container)
-      , type_(type) {}
+      , type_(type) {
+
+    // environment()->register_port(this);
```
[the remark this answers is not in the document]
Comment (revol-xut) [reply include/reactor-cpp/port.hh:49]: because it was a test and is now removed

File: include/reactor-cpp/impl/port_impl.hh [hunk 11]
```diff
@@ -46,6 +66,38 @@ template <class T> auto Port<T>::get() const noexcept -> const ImmutableValuePtr
   return value_ptr_;
 }
 
+template <class T>
+void Port<T>::pull_connection(const ConnectionProperties& properties, const std::vector<BasePort*>& downstream) {
+  Connection<T>* connection = nullptr;
+
+  // TODO: maybe turn this into a switch
+  if (properties.type_ == ConnectionType::Delayed) {
+    connection =                                                                                              // NOLINT
```
[the remark this answers is not in the document]
Comment (revol-xut) [reply include/reactor-cpp/impl/port_impl.hh]: clang-tidy is complaining about gsl::owner<>

File: include/reactor-cpp/impl/port_impl.hh [hunk 12]
```diff
@@ -9,34 +9,54 @@
...
 
-  auto scheduler = environment()->scheduler();
-  this->value_ptr_ = std::move(value_ptr);
-  scheduler->set_port(this);
   this->present_ = true;
+  this->value_ptr_ = std::move(value_ptr);
+
+  this->invoke_set_callback();
+
+  for (auto* const outward : outward_bindings_) {
```
[the remark this answers is not in the document]
Comment (revol-xut) [reply include/reactor-cpp/impl/port_impl.hh:54]: less rounds trips between the port and the enviroment

**Comment by revol-xut** [comment]
Closing because #51 is the small tiny rewrite
